### PR TITLE
Generics support

### PR DESCRIPTION
**Pull request by majakusber**
Revert of PR #1411, adding fixes for parsing edge cases of generics, adding unit tests

### CODE REVIEW

File: regression/cbmc-java/enum1/test.desc
Comment (thk123) [inline regression/cbmc-java/enum1/test.desc]: These hopefully don't need to be turned off any more as it shouldn't crash when they are found.

**Comment by thk123** [comment]
I suggest reviewers look at [this diff](https://github.com/diffblue/cbmc/pull/1434/files/e36d2cd25d121654c22df8c097ad10b017245e45..9b4662d3bc41db55bcca7948b726bcb1b295fd3c) as the original PR (#1322) is reapplied whole sale and has already been reviewed

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: @mgudemann This was to fix parsing classes that extend from generic classes

File: src/java_bytecode/java_types.cpp
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: @mgudemann Deals with reference types that are of the form `T []`

File: src/java_bytecode/java_types.cpp
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: Probably these three cases (`'*'`, `'-'`, `'+'`) should be handled by throwing an exception rather than this ugly partial handling.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [inline src/java_bytecode/java_types.cpp]: @mgudemann this is to resolve when there are multiple bounds on a type

File: unit/java_bytecode/java_bytecode_parse_generics/Bar.java
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/Bar.java]: Nit: this file (and others) are missing a new line at the end of the file (and others). You can make sure CLion inserts these using Settings > Editor > General > Ensure Line Feed at end of file on save

**Review by thk123 — COMMENTED** [review]
Some small nits - I won't approve as I had a hand in writing some of this.

File: src/java_bytecode/java_types.cpp
Comment (chrisr-diffblue) [reply src/java_bytecode/java_types.cpp]: I'd agree with @thk123

File: src/java_bytecode/java_types.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_types.cpp]: I assume these commented out INVARIANT's are intended to be re-instated once the generics support improves - if so, is it worth marking them all with FIXME or TODO comments as well to make that clearer? @thk123 is there any common practice here to include a link to a Jira or PR for in the comment to indicate what the work is covered under?

**Review by chrisr-diffblue — COMMENTED** [review]
Looks pretty good, but like @thk123 a couple if minor nits. Happy to approve if they are addressed, or there's a good explanation why the PR is fine as it is.

File: unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java
Comment (mgudemann) [inline unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:31]: if wildcards can have multiplle bounds, we should have this as test

File: unit/java_bytecode/java_bytecode_parse_generics/GenericFunctions.java
Comment (mgudemann) [inline unit/java_bytecode/java_bytecode_parse_generics/GenericFunctions.java]: maybe add test with two interface bounds and no class, too

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp:203]: maybe give a concrete example here what would be skipped
could this '<' case and the '<' case inside 'L' - object name parsing be merged ?

File: src/java_bytecode/java_types.cpp
Comment (mgudemann) [inline src/java_bytecode/java_types.cpp]: as discussed, this shadows the `class_name` parameter and should be moved down to where this string is used (after the generic signature case)

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp]: For the record - I should say it was me who wrote the  "ugly partial handling" :stuck_out_tongue:

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_convert_class.cpp]: maybe here we should fall back to using the descriptor, and emit a warning, in order to degrade gracefully

**Review by mgudemann — CHANGES_REQUESTED** [review]

File: unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java
Comment (thk123) [reply unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:31]: Sorry comment not clear - wildcards can not have multiple bounds - this java code doesn't compile :)

File: src/java_bytecode/java_bytecode_convert_class.cpp
Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: My understanding is the `descriptor` won't have any generic type information so isn't applicable here (We are only trying to fill out generic information for the class)

Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_class.cpp]: Adding a warning is a good idea however.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp:203]: No - this should only occur for functions that look like:

```java
void <T> foo(T t)
```

etc (i.e. functions that are generic as opposed to part of a generic class). Perhaps a more robust way of dealing with this is to throw and exception and fall back to the descriptor in this case (since this is broadly what this is doing anyway)

**Review by thk123 — CHANGES_REQUESTED** [review]
Blocking until we've resolved issues with the TG repo

**Comment by majakusber** [comment]
Thank you everyone for the reviews, I've now pushed all the changes according to your notes. Let me know what you think.

File: src/java_bytecode/java_types.cpp
Comment (thk123) [reply src/java_bytecode/java_types.cpp:203]: Proposed comment extension:
```C++
// So for example, the following java method:
// static void <T, U> foo(T t, U u, int x);
// Would have a signature that looks like:
// <T:Ljava/lang/Object;U:Ljava/lang/Object;>(TT;TU;I)V
// So we skip all inside the angle brackets and parse the rest of the string:
// (TT;TU;I)V
```

Comment (majakusber) [reply src/java_bytecode/java_types.cpp:203]: Thank you Thomas, adding it in..

**Review by thk123 — APPROVED** [review]

**Comment by thk123** [comment]
Issues with TG repo have been resolved: just need this PRs conflict resolved and diffblue/test-gen#1080 to pass.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: This test case doesn't actually test what is stated in the 'THEN' clause - and I think this functionality has broken. I've just rebased my work on top of this PR, and one of my unit tests is failing because of this. Previously, you could call `is_java_generic_parameter(code.return_type())` and get the answer `true` for this function. I'm not finding this is returning `false` instead, and doing 'method_symbol.show(std::cout)` is showing that the return type for this method is not marked as being generic in any way.

File: src/java_bytecode/java_types.cpp
Comment (NlightNFotis) [inline src/java_bytecode/java_types.cpp:211]: There is a particular problem with this function when passed a `src` string with length `1` only, and it results in a huge recursive call stack, and makes the program result in a `Segmentation Violation` fault. I shall send tests and perhaps a fix in a bit.

**Review by NlightNFotis — CHANGES_REQUESTED** [review]

**Review by mgudemann — APPROVED** [review]
LGTM, please squash some of the commits before merge

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp
Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Thanks for this, this has also been failing the CI tests. The reason is that the return type of the lookup method for symbol table has changed (commit #1444 ) - now it's a reference wrapper around symbolt instead of symbolt.

Comment (chrisr-diffblue) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Yes, that is part of the failure (it's broken my work too :-) ), but my broader point is that this test case doesn't *actually* test what it claims to test - it claims to test that it parses methods with a generic return type, but it only does that partially - it checks that it can parse the method, and that the symbol for the method exists, and is an ID_code object - but it *doesn't* check that the return type of that `code_typet` actually records that it is generic. So, for instance, we could be parsing the the method and then throwing away all the generic information and this test as it is currently written would still be passing. This appears to have happened at some point, because when my code was based off @thk123 branch, the `is_java_generic_parameter(code.return_type())` was correctly returning true for this case, whereas with this branch I'm seeing that the generic information is no longer stored for the return type.

Comment (thk123) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp]: Yes - this is the mess with descriptors of methods being used. As discussed I think before we do [TG-865](https://diffblue.atlassian.net/browse/TG-865) it won't work. The THEN should be updated in a similar way to the TODO directly below.

**Pull request merged**
